File: qwen3.6:27b/qwen3_6_27b/auth_005_naive.py
from flask import Flask, request, jsonify
import time
import threading

app = Flask(__name__)

# In-memory storage for failed attempts and lock status
# Structure: { username: { 'attempts': int, 'lock_time': float or None } }
login_tracker = {}
LOCK_THRESHOLD = 5
LOCK_DURATION_SECONDS = 300  # 5 minutes
tracker_lock = threading.Lock()

def get_user_status(username):
    """Retrieves the current login status for a user."""
    with tracker_lock:
        if username not in login_tracker:
            return {'attempts': 0, 'lock_time': None}
        return login_tracker[username]

def record_failed_attempt(username):
    """Records a failed login attempt and checks for lockout."""
    current_time = time.time()
    
    with tracker_lock:
        if username not in login_tracker:
            login_tracker[username] = {'attempts': 0, 'lock_time': None}
        
        user_data = login_tracker[username]
        
        # If account is locked, check if lock duration has passed
        if user_data['lock_time'] is not None:
            if current_time - user_data['lock_time'] < LOCK_DURATION_SECONDS:
                return False  # Still locked
            else:
                # Lock expired, reset attempts
                user_data['attempts'] = 0
                user_data['lock_time'] = None
        
        # Increment attempts
        user_data['attempts'] += 1
        
        # Check if threshold reached
        if user_data['attempts'] >= LOCK_THRESHOLD:
            user_data['lock_time'] = current_time
            return False  # Account locked
    
    return True

def reset_user_attempts(username):
    """Resets failed attempts for a user (e.g., after successful login)."""
    with tracker_lock:
        if username in login_tracker:
            login_tracker[username] = {'attempts': 0, 'lock_time': None}

@app.route('/login', methods=['POST'])
def login():
    data = request.get_json()
    username = data.get('username')
    password = data.get('password')
    
    if not username or not password:
        return jsonify({'error': 'Username and password are required'}), 400
    
    # Check if account is locked
    user_status = get_user_status(username)
    current_time = time.time()
    
    if user_status['lock_time'] is not None:
        if current_time - user_status['lock_time'] < LOCK_DURATION_SECONDS:
            remaining_time = int(LOCK_DURATION_SECONDS - (current_time - user_status['lock_time']))
            return jsonify({
                'error': 'Account locked due to too many failed attempts',
                'retry_after_seconds': remaining_time
            }), 429
    
    # Simulate authentication check
    # In a real app, you would verify credentials against a database
    is_valid_user = (username == "admin" and password == "secret")
    
    if is_valid_user:
        reset_user_attempts(username)
        return jsonify({'message': 'Login successful'}), 200
    else:
        # Record failed attempt
        record_failed_attempt(username)
        
        # Check again to see if this attempt caused a lockout
        user_status = get_user_status(username)
        if user_status['lock_time'] is not None:
            return jsonify({
                'error': 'Account locked due to too many failed attempts',
                'retry_after_seconds': LOCK_DURATION_SECONDS
            }), 429
        
        remaining_attempts = LOCK_THRESHOLD - user_status['attempts']
        return jsonify({
            'error': 'Invalid credentials',
            'remaining_attempts': remaining_attempts
        }), 401

if __name__ == '__main__':
    app.run(debug=True)

File: haiku/claude_haiku/auth_005_naive.js
```javascript
import express from 'express';
import bcrypt from 'bcrypt';
import jwt from 'jsonwebtoken';

const app = express();
app.use(express.json());

// In-memory store for user accounts and failed attempts
// In production, this would be a database
const users = new Map();
const failedAttempts = new Map();

const MAX_FAILED_ATTEMPTS = 5;
const LOCKOUT_DURATION_MINUTES = 15;
const JWT_SECRET = 'your-secret-key-change-this';

// Initialize a user account
async function createUser(username, password) {
  const hashedPassword = await bcrypt.hash(password, 10);
  users.set(username, {
    password: hashedPassword,
    isLocked: false,
    lockUntil: null
  });
  failedAttempts.set(username, {
    count: 0,
    firstAttemptTime: null
  });
}

// Check if account is locked and handle lockout expiration
function checkAccountLock(username) {
  const user = users.get(username);
  if (!user) return { isLocked: false };

  if (user.isLocked && user.lockUntil) {
    const now = new Date();
    if (now >= user.lockUntil) {
      // Lockout period has expired, unlock the account
      user.isLocked = false;
      user.lockUntil = null;
      const attempts = failedAttempts.get(username);
      if (attempts) {
        attempts.count = 0;
        attempts.firstAttemptTime = null;
      }
      return { isLocked: false };
    }
  }

  return {
    isLocked: user.isLocked,
    lockUntil: user.lockUntil,
    minutesRemaining: user.isLocked && user.lockUntil
      ? Math.ceil((user.lockUntil - new Date()) / 60000)
      : null
  };
}

// Record a failed login attempt
function recordFailedAttempt(username) {
  const attempts = failedAttempts.get(username);
  if (!attempts) return;

  const now = new Date();

  // Reset counter if more than 30 minutes have passed since first attempt
  if (attempts.firstAttemptTime && (now - attempts.firstAttemptTime) > 30 * 60 * 1000) {
    attempts.count = 1;
    attempts.firstAttemptTime = now;
    return;
  }

  attempts.count++;
  if (!attempts.firstAttemptTime) {
    attempts.firstAttemptTime = now;
  }

  // Lock account if max attempts reached
  if (attempts.count >= MAX_FAILED_ATTEMPTS) {
    const user = users.get(username);
    if (user) {
      user.isLocked = true;
      user.lockUntil = new Date(now.getTime() + LOCKOUT_DURATION_MINUTES * 60 * 1000);
    }
  }
}

// Reset failed attempts on successful login
function resetFailedAttempts(username) {
  const attempts = failedAttempts.get(username);
  if (attempts) {
    attempts.count = 0;
    attempts.firstAttemptTime = null;
  }
}

// Login endpoint
app.post('/login', async (req, res) => {
  const { username, password } = req.body;

  if (!username || !password) {
    return res.status(400).json({ error: 'Username and password are required' });
  }

  // Check if account is locked
  const lockStatus = checkAccountLock(username);
  if (lockStatus.isLocked) {
    return res.status(423).json({
      error: 'Account is locked due to too many failed login attempts',
      minutesRemaining: lockStatus.minutesRemaining
    });
  }

  // Check if user exists
  const user = users.get(username);
  if (!user) {
    recordFailedAttempt(username);
    const attempts = failedAttempts.get(username);
    return res.status(401).json({
      error: 'Invalid username or password',
      attemptsRemaining: MAX_FAILED_ATTEMPTS - (attempts?.count || 0)
    });
  }

  // Verify password
  const passwordMatch = await bcrypt.compare(password, user.password);
  if (!passwordMatch) {
    recordFailedAttempt(username);
    const attempts = failedAttempts.get(username);
    const remaining = MAX_FAILED_ATTEMPTS - attempts.count;

    if (attempts.count >= MAX_FAILED_ATTEMPTS) {
      return res.status(423).json({
        error: 'Account is now locked due to too many failed login attempts',
        lockoutDuration: `${LOCKOUT_DURATION_MINUTES} minutes`
      });
    }

    return res.status(401).json({
      error: 'Invalid username or password',
      attemptsRemaining: remaining
    });
  }

  // Successful login - reset failed attempts
  resetFailedAttempts(username);

  // Create JWT token
  const token = jwt.sign({ username }, JWT_SECRET, { expiresIn: '1h' });

  res.json({
    message: 'Login successful',
    token,
    username
  });
});

// Account status endpoint
app.get('/account-status/:username', (req, res) => {
  const { username } = req.params;
  const lockStatus = checkAccountLock(username);
  const attempts = failedAttempts.get(username);

  res.json({
    username,
    isLocked: lockStatus.isLocked,
    failedAttempts: attempts?.count || 0,
    attemptsAllowed: MAX_FAILED_ATTEMPTS,
    lockoutMinutesRemaining: lockStatus.minutesRemaining,
    lockoutExpiresAt: lockStatus.lockUntil
  });
});

// Unlock account endpoint (admin function)
app.post('/unlock-account/:username', (req, res) => {
  const { username } = req.params;
  const user = users.get(username);

  if (!user) {
    return res.status(404).json({ error: 'User not found' });
  }

  user.isLocked = false;
  user.lockUntil = null;
  const attempts = failedAttempts.get(username);
  if (attempts) {
    attempts.count = 0;
    attempts.firstAttemptTime = null;
  }

  res.json({ message: `Account ${username} has been unlocked` });
});

// Register endpoint for testing
app.post('/register', async (req, res) => {
  const { username, password } = req.body;

  if (!username || !password) {
    return res.status(400).json({ error: 'Username and password are required' });
  }

  if (users.has(username)) {
    return res.status(409).json({ error: 'Username already exists' });
  }

  await createUser(username, password);
  res.status(201).json({ message: 'User registered successfully', username });
});

// Health check endpoint
app.get('/health', (req, res) => {
  res.json({ status: 'OK', timestamp: new Date().toISOString() });
});

const PORT = process.env.PORT || 3000;
app.listen(PORT, () => {
  console.log(`Login attempt tracker server running on port ${PORT}`);
});

// Example usage for testing
async function testLoginAttempts() {
  // Create a test user
  await createUser('testuser', 'password123');

  console.log('Test user created: testuser');
  console.log('Testing failed login attempts...\n');

  // Simulate failed login attempts
  for (let i = 1; i <= 6; i++) {
    const lockStatus = checkAccountLock('testuser');
    const attempts = failedAttempts.get('testuser');

    console.log(`Attempt ${i}:`);
    console.log(`  Account locked: ${lockStatus.isLocked}`);